File: solutions/Strings/0003-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<int,int> mp;
        int head=-1,tail=0,ans=0,n=s.length();
        while(tail<n){
            while((head+1<n) && (mp[s[head+1]]==0)){
                head++;
                mp[s[head]]++;
            }
            ans=max(ans,head-tail+1);
            mp[s[tail]]--;
            tail++;
        }
        return ans;
    }
};

File: solutions/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        int l=0;
        int ans=Integer.MIN_VALUE;
        Map<Character,Integer> mp=new HashMap<>();
        for(int r=0;r<n;r++){
            mp.put(s.charAt(r),mp.getOrDefault(s.charAt(r),0)+1);
            while(mp.get(s.charAt(r))>1){
                mp.put(s.charAt(l),mp.get(s.charAt(l))-1);
                l++;
            }
            ans=Integer.max(ans,r-l+1);
        }
        return ans==Integer.MIN_VALUE?0:ans;
    }
}